### PR TITLE
DRILL-6676: Add Union, List and Repeated List types to Result Set Loader

### PR DESCRIPTION
**Pull request by paul-rogers**
Previous commits provided the core "result set loader" (RSL) structure and support for the "mainstream" vector types, including structured types such as maps and lists.

This PR adds the "obscure" (and partly implemented) types used for JSON: (non-repeated) list, repeated list and union.

The union type is complex: it is a bundle of vectors keyed by type, and can accept new types as a run proceeds. A (non-repeated) list is highly complex: it it can act like a repeated list, but with the ability to specify a null state for each entry. The non-repeated List can also act like a union type. This dual/morphing nature of a list required some rather complex magic behind the scenes to support the simple JSON-like interface used by the row set and result set loader mechanisms.

This PR introduces the idea of a "variant" to model unions and non-repeated-lists-as-list-of-unions. The name is taken from Microsoft Basic and simply means a tagged union. (Where "union" is taken from "C".)

Changes include fixing a number of issues with the list vectors, adding support in the column accessors and metadata layers, and adding support for creating vectors from metadata and metadata from vectors.

Unit tests demonstrate how to use the resulting behavior as well as verifying that the behavior is correct.

The focus of this PR is to enable union, list and repeated list support in the RSL and associated mechanisms. It is known that support of these vector types is incomplete: some operators fail when presented with such vectors. It is not the goal here to fix those issues: this is not a PR to fully support these types. Rather, the the scope of this PR is just to the RSL and associated classes.

For more information, see [this wiki entry](https://github.com/paul-rogers/drill/wiki/Batch-Handling-Upgrades).

This PR completes the result set loader work. The next PR in this series will introduce revisions to the scan operator that allow readers to use the RSL. After that, there are revised implementations for the delimited text (e.g. CSV) and JSON readers.

### CODE REVIEW

**Comment by paul-rogers** [comment]
@ilooner, this work is related to your resource management efforts. Please review at your convenience.

**Comment by ilooner** [comment]
Thanks @paul-rogers . I'll run the functional tests and let you know if there are any issues. Since your code is always well documented and tested, I feel comfortable merging this pretty quickly. So maybe we can time-box the review to 2 weeks at most, unless some issue pops up.

@sohami @bitblender since you guys have experience doing low level work with the vectors could you help with looking at this PR? The PR is quite big and would take me a while to go through by myself.

File: exec/vector/src/main/java/org/apache/drill/exec/vector/complex/ListVector.java [hunk 1]
```diff
@@ -250,22 +270,140 @@ public void load(UserBitShared.SerializedField metadata, DrillBuf buffer) {
...
+          typeMutator.setSafe(i, (bitsAccessor.get(i) == 0)
+              ? UnionVector.NULL_MARKER
+              : prevType);
+        }
+      } else {
+
+        // The value is not nullable. (Perhaps it is a map.)
+        // Note that the original design of lists have a flaw: if the sole member
+        // is a map, then map entries can't be nullable when the only type, but
+        // become nullable when in a union. What a mess...
```
Comment (ilooner) [inline exec/vector/src/main/java/org/apache/drill/exec/vector/complex/ListVector.java]: I'm a little confused. Is this still a problem with this new code, or did you fix it? If there is still a pre-existing issue could you make a jira describing it so that we have a record incase we bump into it later.

File: exec/vector/src/main/java/org/apache/drill/exec/vector/complex/ListVector.java [hunk 2]
```diff
@@ -250,22 +270,140 @@ public void load(UserBitShared.SerializedField metadata, DrillBuf buffer) {
...
+  }
+
+  private UnionVector createUnion() {
+    MaterializedField newField = MaterializedField.create(UNION_VECTOR_NAME, Types.optional(MinorType.UNION));
+    UnionVector unionVector = new UnionVector(newField, allocator, null);
+
+    // For efficiency, should not create a reader that will never be used.
+    // Keeping for backward compatibility.
+
+    reader = new UnionListReader(this);
```
Comment (ilooner) [inline exec/vector/src/main/java/org/apache/drill/exec/vector/complex/ListVector.java]: Why is this needed for backward compatibility?

**Comment by ilooner** [comment]
@paul-rogers I added some more checkstyle checks recently, and they fail when the PR is rebased onto the latest master. Could you rebase and fix the few checkstyle errors?

**Comment by ppadma** [comment]
@paul-rogers @ilooner I can review this PR. It will take some time. I will post comments as I make progress.  meanwhile, @ilooner, can you run functional tests ?

**Comment by paul-rogers** [comment]
@ilooner, thanks much for taking a look at this PR. Agree it would be great for @ppadma to review this as she's been following allow with earlier work in this series.

The code here can be overwhelming if you dive right in. The best way to start is to look at the tests: to see how the code is intended to be used. The API is intended to be simple; the large bulk of code is needed to create that simple API by hiding the highly complex stuff that must be done to keep multiple vectors in sync -- with ever increasing complexity as you move from fixed-width to nullable to variable width to arrays to maps to unions to lists.

Rebased on latest master. Fixed check style issues from latest rules.

File: exec/vector/src/main/java/org/apache/drill/exec/vector/complex/ListVector.java (hunk 1 above)
Comment (paul-rogers) [reply exec/vector/src/main/java/org/apache/drill/exec/vector/complex/ListVector.java]: The first thing to understand is that union vectors are broken in most operators, and so major bugs remain in union vectors that have not had to be fixed. This code fixes or works around some of the bugs.

Unions have a null bit for the union itself. That is, a union can be an int, say, or. a Varchar, or null. Oddly, the Int and Varchar can also be null (we use nullable vectors so we can mark the unused values as null.) So, we have a two-level null bit. The most logical way to interpret it is that a union value can be:

* Untyped null (if the type is not set and the null bit (really, the isSet bit) is unset.)
* Typed null if the type is set and EITHER the union's isSet bit is unset OR the union's isSet bit is set, but the data vector's isSet bit is not set.

It is not clear in the code which convention is assumed, or if different code does it differently.

Now, add all that to a list. A list can be a list of something (ints, say, or maps.) When the list is a list of maps, the entire value for a row can be null. But individual maps can't be null. In a list, however, individual ints can be null (because we use a nullable int vector.)

So, when a list of (non-nullable maps) converts to a list of unions (one of which is a map), we suddenly now have the list null bit and the union null bit to worry about. We have to go and back-patch the isSet vector for all the existing map entries in the new union so that we don't end up with all previous entries becoming null by default.

See why the code notes that this is a mess?

And, it is hard to file a bug because this is a design problem. If the list and union vectors don't need to work (they barely work today), then any design is fine.

See DRILL-5955, DRILL-6037, DRILL-5958, DRILL-6046, DRILL-6062, DRILL-3806 to get a flavor for the issues.

Fundamental issue: should Drill support unions and lists? Is the current approach compatible with SQL? Is there a better approach?

File: exec/vector/src/main/java/org/apache/drill/exec/vector/complex/ListVector.java (hunk 2 above)
Comment (paul-rogers) [reply exec/vector/src/main/java/org/apache/drill/exec/vector/complex/ListVector.java]: This vector already creates a reader for every vector. Since I didn't have time to track down uses, I kept this code. The Union vector is normally created in the older "complex writer" which created the reader.

Moving forward, we should revisit union vectors (and the older complex writer). At that time, we can determine if the vector needs to own a reader, or if the reader can be created as needed.

If we create a "result set reader" that works cross-batch in parallel with the result set loader (for writing), then vectors would not carry their own readers.

Yes indeed, union vectors are a mess.

**Comment by ilooner** [comment]
Thanks for the help @ppadma. Ran the tests. There are a few failures but those are known issues unrelated to your change. So I think we are all good with respect to the build.

**Comment by paul-rogers** [comment]
Rebased on latest master.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/rowSet/impl/BuildFromSchema.java [hunk 3]
```diff
@@ -96,4 +164,89 @@ private void expandMap(ObjectWriter colWriter, ColumnMetadata colSchema) {
...
+   * build writer for the components of the column. We do that recursively
+   * here.
+   *
+   * @param colWriter the writer for the (possibly structured) column
+   * @param colSchema the schema definition for the column
+   */
+
+  private void expandColumn(ObjectWriter colWriter, ColumnMetadata colSchema) {
+
+    if (colSchema.structureType() == StructureType.MULTI_ARRAY) {
```
[the remark this answers is not in the document]
Comment (paul-rogers) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/rowSet/impl/BuildFromSchema.java]: Fixed. Moved the method into the `ColumnMetadata` interface where it belongs.

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/rowSet/impl/TestResultSetLoaderUnions.java [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+        .add("id", MinorType.INT)
+        .addList("list")
+          .resumeSchema()
+        .buildSchema();
+
+    try {
+      schema.metadata("list").variantSchema().becomeSimple();
+    } catch (final IllegalStateException e) {
+      // expected
+    }
```
[the remark this answers is not in the document]
Comment (paul-rogers) [reply exec/java-exec/src/test/java/org/apache/drill/exec/physical/rowSet/impl/TestResultSetLoaderUnions.java]: Fixed.

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/rowSet/impl/TestResultSetLoaderUnions.java [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+          .addType(MinorType.VARCHAR)
+          .addType(MinorType.INT)
+          .resumeSchema()
+        .buildSchema();
+
+    try {
+      schema.metadata("list").variantSchema().becomeSimple();
+    } catch (final IllegalStateException e) {
+      // expected
+    }
```
[the remark this answers is not in the document]
Comment (paul-rogers) [reply exec/java-exec/src/test/java/org/apache/drill/exec/physical/rowSet/impl/TestResultSetLoaderUnions.java]: Fixed.

File: exec/vector/src/main/java/org/apache/drill/exec/vector/complex/BaseRepeatedValueVector.java [hunk 6]
```diff
@@ -212,6 +212,18 @@ protected void replaceDataVector(ValueVector v) {
...
   }
 
+  public void setChildVector(ValueVector childVector) {
+
+    // When created, the list uses the default vector of type LATE.
+    // That entry appears as a child vector. Remove it and add the
+    // new type instead.
+
+    assert vector == DEFAULT_DATA_VECTOR;
+    replaceDataVector(childVector);
```
[the remark this answers is not in the document]
Comment (paul-rogers) [reply exec/vector/src/main/java/org/apache/drill/exec/vector/complex/BaseRepeatedValueVector.java]: Done.

**Comment by paul-rogers** [comment]
Thanks @ppadma for the review!

Addressed comments. Rebased on latest master. Reran unit tests.

**Comment by arina-ielchiieva** [comment]
@paul-rogers please squash commits into one.

**Comment by paul-rogers** [comment]
@arina-ielchiieva, squashed commits. Rebased on latest master.

**Comment by sohami** [comment]
@paul-rogers - Please resolve the conflicts.

**Comment by paul-rogers** [comment]
@sohami, rebased on latest master. Resolved merge conflicts. Resolved issues due to new check-style rules. Let's please get this PR committed before we hit more merge conflicts.
